### PR TITLE
fix modz bug

### LINKED ISSUE #162: MODZ drops replicate groups containing missing values
Running `consensus(operation="modz")` will silently drop groups (as specified in the argument `replicate_columns`) if one of the groups contains `NA` in one of the `replicate_columns`.

The correct behavior would be to consider NA as a valid group member, or, at the very least, warn the user that MODZ is invalid for groups containing missing values.

### PR DESCRIPTION
**Pull request by axiomcura**
# Description
in this PR, `modz` is updated so replicate groups with missing values in `replicate_columns` are no longer silently dropped, and MODZ handles edge cases where correlation-based weights become undefined.

The changes are intentionally small and targeted:
- `modz()` now uses `groupby(..., dropna=False)` so groups containing `NA` metadata are preserved during consensus generation.

This PR also adds tests in `tests/test_cyto_utils/test_modz.py` for:
- replicate groups containing `NA` values
- profiles with zero-valued features that previously caused invalid weighting behavior

No existing tests were changed, and the overall structure of the MODZ implementation remains the same to keep the fix small and ready for review.

This PR closes #162 and #170
## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved weight normalization with safeguards against division-by-zero errors
  * Enhanced handling of missing data in replicate groups—now retained instead of excluded

* **Tests**
  * Added test coverage for NaN values in replicate group preservation
  * Added test coverage for zero-value profiles in consensus computation

### CODE REVIEW

File: pycytominer/cyto_utils/modz.py [hunk 1]
```diff
@@ -73,13 +73,17 @@ def modz_base(
     cor_df = cor_df.clip(lower=0)
 
     # Get average correlation for each profile (will ignore NaN)
-    raw_weights = cor_df.mean(axis=1)
+    raw_weights = cor_df.mean(axis=1).fillna(min_weight)
```
Comment (d33bs) [inline pycytominer/cyto_utils/modz.py]: This feels like a larger implied change. Consider providing documentation to help ground this decision somehow.

File: pycytominer/cyto_utils/modz.py [hunk 2]
```diff
@@ -73,13 +73,17 @@ def modz_base(
...
 
     # Threshold weights (any value < min_weight will become min_weight)
     raw_weights = raw_weights.clip(lower=min_weight)
 
     # normalize raw_weights so that they add to 1
-    weights = raw_weights / sum(raw_weights)
+    weight_sum = raw_weights.sum()
+    if weight_sum == 0:
```
Comment (d33bs) [inline pycytominer/cyto_utils/modz.py:83]: Do we test for negative cases too somehow?

File: tests/test_cyto_utils/test_modz.py [hunk 3]
```diff
@@ -191,3 +192,51 @@ def test_modz_unbalanced_sample_numbers():
         },
     )
     pd.testing.assert_frame_equal(expected_result, consensus_df)
+
+
+def test_modz_preserves_na_replicate_groups():
```
Comment (d33bs) [inline tests/test_cyto_utils/test_modz.py:197]: Have we tested for multi-column replicate groups?

Comment (d33bs) [inline tests/test_cyto_utils/test_modz.py:197]: Consider providing docstrings for both this and the other test to ensure we describe what the goal of the test is.

**Comment by d33bs** [comment]
Thanks @axiomcura for the changes here! I left some comments but wasn't as familiar with the functionality and history here. @gwaybio, could I ask you to take a look as well?

File: pycytominer/cyto_utils/modz.py (hunk 1 above)
Comment (gwaybio) [reply pycytominer/cyto_utils/modz.py]: line 75 also needs to change - what is the justification to fill NAs with min weight?

File: pycytominer/cyto_utils/modz.py (hunk 2 above)
Comment (gwaybio) [reply pycytominer/cyto_utils/modz.py:83]: please also add a comment here - what is this edge case? why is it important to prevent?

File: tests/test_cyto_utils/test_modz.py (hunk 3 above)
Comment (gwaybio) [reply tests/test_cyto_utils/test_modz.py:197]: agree - all tests need docs like this

**Review by gwaybio — CHANGES_REQUESTED** [review]
A few comments, which may trigger additional comments once resolved. I also don't see a test where `min_weight` is triggered

File: tests/test_cyto_utils/test_modz.py (hunk 3 above)
Comment (axiomcura) [reply tests/test_cyto_utils/test_modz.py:197]: tests added!

Comment (axiomcura) [reply tests/test_cyto_utils/test_modz.py:197]: I think there's a test that checks for this: https://github.com/cytomining/pycytominer/blob/341cbe603abb9125149802a7eae793df36b159cd/tests/test_cyto_utils/test_modz.py#L82-L110

File: pycytominer/cyto_utils/modz.py (hunk 2 above)
Comment (axiomcura) [reply pycytominer/cyto_utils/modz.py:83]: I noticed that there wasn’t a good way to handle the case where all raw weights are `0`, which can generate a series of `NaN` weights during normalization. This can happen when all replicate correlations are non-positive and `min_weight=0`; falling back to equal weights prevents invalid consensus values while still producing a reasonable profile.

Another option is to just raise an error for this edge case were all weight are 0. 

A 3rd option is just to remove this check c: (if this edge cases is likely to never happen) 

LMK what you think!

Comment (axiomcura) [reply pycytominer/cyto_utils/modz.py:83]: also updated with comments !

File: pycytominer/cyto_utils/modz.py (hunk 1 above)
Comment (axiomcura) [reply pycytominer/cyto_utils/modz.py]: This was a mistake on my end! now removed!  (why I didn't change the document) code that I forgot to remove

**Review by gwaybio — APPROVED** [review]

**Comment by axiomcura** [comment]
@d33bs LMK if I have answered all your comments and suggestions before I merge!

**Comment by axiomcura** [comment]
@d33bs Greg approved the changes, just wanted to know if you had any comments, if not, I'd be happy to merge c:

File: pycytominer/cyto_utils/modz.py (hunk 2 above)
Comment (d33bs) [reply pycytominer/cyto_utils/modz.py:83]: What would happen for a -1 min weight / weight sum? I'm not sure of the circumstances which would make this possible, but bringing it up because it seems technically an option.

File: pycytominer/cyto_utils/modz.py [hunk 4]
```diff
@@ -151,7 +157,7 @@ def modz(
 
     modz_df = (
         population_df
-        .groupby(replicate_columns)
+        .groupby(replicate_columns, dropna=False)
```
Comment (d33bs) [inline pycytominer/cyto_utils/modz.py:160]: Would it be worth including optionality here in case dropna is undesired?

**Review by d33bs — APPROVED** [review]
Generally looking good to me, left a couple more comments.

File: pycytominer/cyto_utils/modz.py (hunk 4 above)
Comment (gwaybio) [reply pycytominer/cyto_utils/modz.py:160]: for this, i think better for us to be opinionated

**Comment by gwaybio** [comment]
thanks @axiomcura - i'm merging this in!
